### PR TITLE
Closes Nomad UI showing 'vUnknown' #10672

### PR DESCRIPTION
**Pull request by JBhagat841**
This PR closes [10672](https://github.com/hashicorp/nomad/issues/10672).

The the Nomad UI previously made a request to the `agent/self` endpoint and returned information about the Nomad Agent Version via the Nomad Agent Version.

This PR edits that to retrieve the Nomad Agent Version from the Nomad Agent Config which is created at build time.

The rationale behind this change was that, we have a guaranteed result for Nomad Agent Config Version. Whereas, we didn't have a guarantee that we'd always have a Nomad Agent Member because Clients are not included in the Serf Gossip Protocol.

We know the bug always occurs when we have clients, but no servers.

This PR does not resolve the issue pointed out by @DingoEatingFuzz regarding [cached requests](https://github.com/hashicorp/nomad/issues/6458). That would require us to dynamically serve the `index.html` and also calculate the Nomad Agent Version based on text files, we believe. We will eventually handle this issue when we tackle this [issue](https://app.asana.com/0/758378679700210/1165474804795867/f).

### CODE REVIEW

File: ui/app/templates/components/gutter-menu.hbs [hunk 1]
```diff
@@ -68,9 +68,11 @@
         <li><LinkTo @route="topology" @activeClass="is-active" data-test-gutter-link="topology">Topology</LinkTo></li>
       </ul>
     </aside>
+    {{#if this.system.agent.version}}
```
Comment (JBhagat841) [inline ui/app/templates/components/gutter-menu.hbs:71]: If we're type guarding `agent.config` in `system.js` just in case there's an API change where a case like `config` becomes `Config` (uppercasing because of how Go works then maybe we should add the Elvis operator here.

`this.system.agent?.version`

Comment (JBhagat841) [reply ui/app/templates/components/gutter-menu.hbs:71]: This is also an implicit design decision to say if we don't have the `agent.version` then we won't render the Agent Version on the screen.

How do we approach this on the Hashicorp team?

File: ui/mirage/factories/agent.js [hunk 2]
```diff
@@ -8,20 +8,51 @@ const AGENT_STATUSES = ['alive', 'leaving', 'left', 'failed'];
 
 export default Factory.extend({
   id: i => (i / 100 >= 1 ? `${UUIDS[i]}-${i}` : UUIDS[i]),
-  name: () => `nomad@${faker.random.boolean() ? faker.internet.ip() : faker.internet.ipv6()}`,
+  config: makeConfig,
```
Comment (JBhagat841) [inline ui/mirage/factories/agent.js:11]: Is the correct pattern here to have a function using the `make` prefix?

File: ui/mirage/factories/agent.js [hunk 3]
```diff
@@ -8,20 +8,51 @@ const AGENT_STATUSES = ['alive', 'leaving', 'left', 'failed'];
...
-      dc: faker.helpers.randomize(DATACENTERS),
+      name: generateName(),
+      status: faker.helpers.randomize(AGENT_STATUSES),
+      serfPort: faker.random.number({ min: 4000, max: 4999 }),
     };
   },
+
+  afterCreate(agent) {
+    const address = generateAddress.bind(agent);
+    const tags = generateTags.bind(agent);
```
Comment (JBhagat841) [inline ui/mirage/factories/agent.js:23]: These are computed properties based on other properties of the member object. Unsure if `afterCreate` was the best option. My next best idea was adding some init function to the member property and then doing something like this. I saw it on StackOverflow but then that pattern wasn't used in the codebase and afterCreate seemed like better documented pattern in Ember Mirage.

```
{
  x: string,
  y: boolean,
  init: () => this.y ? this.x : 'haha yeah'
}.init();
```

File: ui/mirage/serializers/agent.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import ApplicationSerializer from './application';
+
+export default ApplicationSerializer.extend({
+  keyForAttribute(str) {
+    if (str === 'config' || str === 'member') {
+      return str;
+    }
+    return ApplicationSerializer.prototype.keyForAttribute.apply(this, arguments);
+  },
```
Comment (JBhagat841) [inline ui/mirage/serializers/agent.js:9]: The alternative here could be turning `keyCase` on `ApplicationSerializer` into a more generic API.

```
const keyCase = (str, config, customConfig) => {
  switch (config.type):
    case overridePascalCase:
      return str;
    case customConfig:
      return CustomConfig();
    default:
      return defaultLogic();
}
```

Seems like overkill, but I think its still a worthwhile thought experiment to how we think about DRY v WET, readability, and standards.

File: ui/mirage/factories/agent.js [hunk 5]
```diff
@@ -8,20 +8,51 @@ const AGENT_STATUSES = ['alive', 'leaving', 'left', 'failed'];
...
+      member: {
+        ...agent.member,
+        address: address(),
+        tags: tags(),
+      },
+    });
+  },
 });
+
+function generateName() {
```
Comment (JBhagat841) [inline ui/mirage/factories/agent.js:34]: kept these down here and ignored ES6 for the benefit of hoisting to make the file easier to read, do we have opinions on this?

File: ui/mirage/factories/agent.js [hunk 6]
```diff
@@ -8,20 +8,51 @@ const AGENT_STATUSES = ['alive', 'leaving', 'left', 'failed'];
...
-  status: () => faker.helpers.randomize(AGENT_STATUSES),
-  serfPort: () => faker.random.number({ min: 4000, max: 4999 }),
-
-  address() {
-    return this.name.split('@')[1];
-  },
-
-  tags() {
-    const rpcPortCandidate = faker.random.number({ min: 4000, max: 4999 });
+  member() {
```
Comment (JBhagat841) [inline ui/mirage/factories/agent.js:13]: is there anything stopping us from simplifying this to:

```
member: {
  x,
  y,
  z
}
```

File: ui/mirage/factories/agent.js (hunk 2 above)
Comment (lgfa29) [reply ui/mirage/factories/agent.js:11]: Looking at similar factories (like [node](https://github.com/hashicorp/nomad/blob/main/ui/mirage/factories/node.js) or [task-group](https://github.com/hashicorp/nomad/blob/main/ui/mirage/factories/task-group.js)) it seems like the `make` pattern is used when the resulting object is dynamic, which doesn't seem to be the case here since we are hard coding a version.

So I think we can just have this as:

```js
export default Factory.extend({
  id: i => (i / 100 >= 1 ? `${UUIDS[i]}-${i}` : UUIDS[i]),

  config: {
    Version: {
      Version: '1.1.0',
      VersionMetadata: 'ent',
      VersionPrerelease: 'dev',
    },
  },
// ...
```

File: ui/mirage/factories/agent.js (hunk 6 above)
Comment (lgfa29) [reply ui/mirage/factories/agent.js:13]: This attribute has dynamic values, so I guess that's why it needs to be a function. Although I tried it as an object and it worked as well, so I guess it's up to you 😄

File: ui/mirage/factories/agent.js (hunk 3 above)
Comment (lgfa29) [reply ui/mirage/factories/agent.js:23]: Why did these have to be moved to an `afterCreate`?  According to the [docs](https://www.ember-cli-mirage.com/versions/v0.4.x/factories/) you can reference other attributes dynamically, and that's how it was implemented before.

File: ui/mirage/factories/agent.js (hunk 5 above)
Comment (lgfa29) [reply ui/mirage/factories/agent.js:34]: I don't have any personal preference, though them seem to be called in line in other factories? https://github.com/hashicorp/nomad/blob/main/ui/mirage/factories/task-group.js#L9

File: ui/mirage/serializers/agent.js (hunk 4 above)
Comment (lgfa29) [reply ui/mirage/serializers/agent.js:9]: `keyCase` is used by every other factory, I'm not sure there would be much benefit in modifying it just for this case. The way it's implemented right now, we're using inheritance to expand on a default parent behaviour, which seems reasonable to me?

**Review by lgfa29 — CHANGES_REQUESTED** [review]
I noticed an issue in the http://localhost:4200/ui/clients and http://localhost:4200/ui/servers pages when using Mirage:
image

I think it's because we removed `id` from `member` 😅

File: ui/mirage/factories/agent.js (hunk 3 above)
Comment (JBhagat841) [reply ui/mirage/factories/agent.js:23]: I still don't know the answer about why. I believe its because the reference to `this`. Let me get you a better answer later.

Comment (johncowen) [reply ui/mirage/factories/agent.js:23]: I need to have a better look at this, but right now I'm not sure what the reason is for binding. You could just pass the `agent` data in as an argument instead of binding and using `this`. Someone might look at those functions and think, "yeah ok, but what is `this` here", using `generateName(agent)` comments the code as you go so I know that the data being read is something to do with `agent`, `this` doesn't give me that context.

Also, at first glance, you only read one piece of data of the `agent` data in each function, you could just pass the single piece of data in, or even maybe (again I'd have to look closer) just do it inline.

All very stylistic this and not super important I know, but I think `bind` usage is uncommon overall in project, and trying to channel other engs on the project I would guess they suggest not using it here - but I could be wrong.

I'll look a little deeper in a sec, but hope that makes sense.

File: ui/app/services/system.js [hunk 7]
```diff
@@ -39,7 +39,12 @@ export default class SystemService extends Service {
         .authorizedRawRequest(`/${namespace}/agent/self`)
         .then(jsonWithDefault({}))
         .then(agent => {
-          agent.version = agent.member?.Tags?.build || 'Unknown';
+          if (agent.config.Version) {
+            const { Version, VersionPrerelease, VersionMetadata } = agent.config.Version;
+            agent.version = Version;
+            if (VersionPrerelease) agent.version += `-${VersionPrerelease}`;
+            if (VersionMetadata) agent.version += `+${VersionMetadata}`;
```
Comment (johncowen) [inline ui/app/services/system.js:46]: Little quibble/nit here, maybe use 

```javascript
agent.version = `${agent.version}-${VersionPrerelease}`
```

etc if you are using template literals anyway?

To me it reads a little easier and personally I like using template strings instead of using `+`/`+=` for string concatenation. Complete nit though, so I'll leave it up to you.

**Review by johncowen — COMMENTED** [review]
Hey @JBhagat841 :wave:

So I saw you had Luiz help out here, but as I promised I'd look at this for you, I've added a few inline comments here.

Overall super good! Love that you gave a nice description so I knew what to expect, and the more in-depth inline comments you left let me know what you were thinking/looking for feedback on. Really helpful, so thankyou!

I just took a brief scan over it and spotted a couple of ember-like things that a non-ember person might not be familiar with, and I was gonna give it a test drive and a deeper look, which I can still do if you like? Let me know, otherwise I can leave you folks to it 👍

File: ui/app/services/system.js (hunk 7 above)
Comment (JBhagat841) [reply ui/app/services/system.js:46]: We're conditionally adding to `agent.version` because `Prerelease` and `Metadata` can be empty strings.

So would the better way to do this be:

```
agent.version = Version;
if (VersionPrerelease) {
  agent.version = `${agent.version}-${VersionPrerelease}`
}
if (VersionMetadata) {
  agent.version = `${agent.version}-${MetaData}`
}
```

File: ui/mirage/serializers/agent.js (hunk 4 above)
Comment (JBhagat841) [reply ui/mirage/serializers/agent.js:9]: Gotcha. The other factories that need to further abstract against the `keyCase` logic also follow this pattern so I see that its inline with what we're currently doing.

File: ui/mirage/factories/agent.js (hunk 5 above)
Comment (JBhagat841) [reply ui/mirage/factories/agent.js:38]: These are dynamic attributes which we would normally put in function calls.

tags() {
  return `some${logic}`;
}

However, we're inside of an object, and we no longer get the benefit of automatically running the function when the new instance of this Factory is created.

To me, normally in this case we wouldn't have an object or a list for a property because that would indicate a relationship to another entity, hence why we'd use `trait` or `afterCreate`. I think the pattern here would be consistent because we'd have to use this pattern for `tags` since it's really verbose.

File: ui/app/templates/components/gutter-menu.hbs (hunk 1 above)
Comment (JBhagat841) [reply ui/app/templates/components/gutter-menu.hbs:71]: Ah... that's pretty odd. So we're not actually using JavaScript in `hbs` files. I didn't pick up on not having that option coming from React where we can use JavaScript in `jsx`.

File: ui/mirage/factories/agent.js (hunk 2 above)
Comment (JBhagat841) [reply ui/mirage/factories/agent.js]: Great call out.

File: ui/mirage/factories/agent.js (hunk 3 above)
Comment (JBhagat841) [reply ui/mirage/factories/agent.js]: This situation is a little different since, we're not working with attributes of the class. In order to access `name` we'd have to access `this.member.name` which is a circular reference.

I think we'd normally have `Member` be its own entity in a database that's related to `Agent` and we wouldn't be writing it as an object here. And the docs do mention that when you're working with a relationship that you should use `trait` or `afterCreate`.

I'm not sure how else we can get around the `this` reference issue.

**Comment by JBhagat841** [comment]
> I noticed an issue in the http://localhost:4200/ui/clients and http://localhost:4200/ui/servers pages when using Mirage:
> image
> 
> I think it's because we removed `id` from `member` 😅

I missed this comment, maybe `Member` should be it own entity.

**Review by johncowen — COMMENTED** [review]
Hey so I gave it a little test drive and all looks good to me (apart from the clients/servers pages that I think you are looking at already so I'll leave that with you), I learnt some also so 🎉

**Review by DerekStrickland — APPROVED** [review]
Nothing blocking from me. Just reading and learning.

File: ui/mirage/factories/agent.js [hunk 8]
```diff
@@ -8,20 +8,40 @@ const AGENT_STATUSES = ['alive', 'leaving', 'left', 'failed'];
...
-  tags() {
-    const rpcPortCandidate = faker.random.number({ min: 4000, max: 4999 });
+  member: () => {
+    const name = generateName();
+    const serfPort = faker.random.number({ min: 4000, max: 4999 });
     return {
-      port: rpcPortCandidate === this.serfPort ? rpcPortCandidate + 1 : rpcPortCandidate,
-      dc: faker.helpers.randomize(DATACENTERS),
+      Name: name,
+      SerfPort: serfPort,
```
Comment (lgfa29) [inline ui/mirage/factories/agent.js:25]: Just this be just `Port` to match the API result?

```json
  "member": {
    "Addr": "172.31.14.218",
    "DelegateCur": 4,
    "DelegateMax": 5,
    "DelegateMin": 2,
    "Name": "ip-172-31-14-218.global",
    "Port": 4648,
    "ProtocolCur": 2,
    "ProtocolMax": 5,
    "ProtocolMin": 1,
    "Status": "alive",
    "Tags": {
      "dc": "dc1",
      "mvn": "1",
      "bootstrap": "1",
      "id": "a9a68da9-0d4e-c9fa-6774-4017cfdae40d",
      "port": "4647",
      "expect": "1",
      "rpc_addr": "172.31.14.218",
      "role": "nomad",
      "region": "global",
      "vsn": "1",
      "build": "0.12.12",
      "raft_vsn": "2"
    }
  },
```

**Review by lgfa29 — APPROVED** [review]
Minor comments, but it LGTM. 

Nice work 🎉

File: ui/mirage/config.js [hunk 9]
```diff
@@ -335,14 +335,12 @@ export default function() {
...
-      Members: this.serialize(agents.all()),
+      Members: this.serialize(agents.all()).map(({ member }) => ({ ...member })),
     };
   });
 
   this.get('/agent/self', function({ agents }) {
-    return {
-      member: this.serialize(agents.first()),
-    };
+    return agents.first();
```
[the remark this answers is not in the document]
Comment (JBhagat841) [reply ui/mirage/config.js:343]: Serialization happens by default. Before this value is returned Mirage is checking is there is a custom Serializer if not it'll apply the behavior in ApplicationSerializer.

We should serialize directly within the route if we're looking to mutate the entity or list of entities because we're working with a collection of ModelClasses that don't have access to methods like `map`, `reduce` and `filter`.

tl;dr - We serialize when we need to use Array and Object Methods before returning out of the route like we should in `agent/members`. The default behavior is to serialize the Collection before returning it and we get that behavior for free because Mirage is already doing this by default.

File: ui/mirage/factories/agent.js (hunk 8 above)
Comment (JBhagat841) [reply ui/mirage/factories/agent.js:25]: Great callout.

File: ui/mirage/config.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (lgfa29) [reply ui/mirage/config.js:343]: Ah got it, thanks!
